### PR TITLE
Respect rich text features when converting between db and editor HTML

### LINKED ISSUE #3852: Specifying rich text features does not whitelist only these features
Starting Wagtail 1.12, we can specify the features allowed in a rich text field, for example just bold and italic.

Unfortunately, there is no server side validation for that. I could work around that by overriding the `clean` method of a given model and by defining a subclass of `DbWhitelister` that contains a subset of the `element_rules` dictionary.

So if you copy/paste content from another field or document, it keeps links etc even if you didn’t allow that feature.

A solution for that could be to define a dictionary linking `features` to `element_rules`. Then in the `clean` method of `RichTextField` or block, we could define a new instance of `DbWhitelister` that gets the class attribute `element_rules` overridden by a homonym instance class containing a subset of `element_rules`.

### PR DESCRIPTION
**Pull request by gasman**
Fixes #3852; incorporates #4060

The `construct_whitelister_element_rules` hook is deprecated, and `register_rich_text_embed_handler` / `register_rich_text_linktype_handler` are dropped (being previously undocumented APIs), in favour of defining them through rich text features.

`expand_db_html` (which does the regexp-based translation from database representation into editor or front-end HTML) has been refactored into a new `wagtail.core.rich_text.rewriters` module to make things more pluggable and hopefully (as a future development) permit fancier kinds of rewriting, such as annotating documents with file size. For now, it means we can split up the dbHTML->frontend and dbHTML->editor workflows...

Finally, the dbHTML->editor and editor->dbHTML (aka whitelisting) conversions have been wrapped into a single rich-text-feature-aware `EditorHTMLConverter` class. (The plan for Draftail integration is for the ContentState heavy lifting to be contained in an equivalent ContentStateConverter...)

### CODE REVIEW

**Comment by BertrandBordage** [comment]
Thanks for this very important work @gasman!

Before reviewing in detail, I have a first question: why not use [bleach](https://github.com/mozilla/bleach) instead of writing our own HTML whitelister? I am the first one against adding dependencies, but for a sensible feature like this, we want a secure solution. Bleach has the advantage of containing a lot of security checks to avoid JavaScript injection. It also allows attributes whitelisting, which could be a nice addition to our whitelister.

I know this is for a back office and we assume editors are not malicious. But this is not enough for security, because of these scenarii:
- Some sites have untrusted editors. For example, MadeWithWagtail has a few editors, and we don’t really know if we can trust some of them. We don’t want a restricted user to make a permission escalation using JavaScript to get an admin session key.
- Users don’t have to be malicious in order to do malicious actions. Two examples:
  - a computer infected with a virus automatically adding multiple JavaScript injection types to all HTML fields that send to a server all session cookies
  - a user copy/pasting HTML from a web page or an HTML mail that contain some JavaScript injection.

Other improvements of using bleach: it’s using html5lib and not BeautifulSoup, which has these consequences:
- html5lib correctly handles HTML5, which is not the case of `html.parser` that you use here. Despite we find this absolutely stupid or not, this is valid HTML5:
    ```HTML
    <p>a<p>c<p>d</p>
    <ul>
      <li>a
      <li>b
    </ul>
    ```
- It should be faster since BeautifulSoup is unfortunately pretty badly optimized.
- We could drop the BeautifulSoup dependency, which is significantly bigger than bleach.
  I know BeautifulSoup is also used to compare revisions, but most of this comparison code could be removed. For some reason, [difflib.HtmlDiff](https://docs.python.org/3/library/difflib.html?highlight=htmldiff#difflib.HtmlDiff) was not used, but I don’t see why.

**Comment by BertrandBordage** [comment]
About `difflib.HtmlDiff`, I think I know why: we don’t use it because it generates a table.
What I wanted to mention is in fact [lxml.html.diff.htmldiff](http://lxml.de/lxmlhtml.html#html-diff) that works very well without striping tags.

For our revision comparison, we can just strip tags using [strip_tags](https://docs.djangoproject.com/en/1.11/ref/utils/#django.utils.html.strip_tags) from Django instead of BeautifulSoup.

**Comment by gasman** [comment]
Hi @BertrandBordage,
Our whitelister does support whitelisting of attributes, and [does use html5lib as its backend](https://github.com/gasman/wagtail/blob/e6a115f6775b08e1eb7fa16a7fa30015de790ee1/wagtail/core/whitelist.py#L100) - I think the only place I've used `html.parser` is in tests. 

(To be honest, I'd be very happy to drop our dependency on html5lib, given the chance... it's caused installation issues like #3660, and the whole "0.999999999" business means that we can't use version numbers as an indication of stability. It's now at the point where I'm just telling people "use the six-nines version - it seems to be the least broken one somehow..." :-) )

The reason I built the whitelister from scratch is that it also handles the conversion of links and embeds - for embeds in particular, we need to be able to match any element with a `data-embedtype` attribute and process it with a rule inserted via hooks. I don't think bleach provides that level of customisability.

In any case, this was a decision made in the early days of Wagtail, and I don't think it makes sense to rewrite it now, given: 1) the `construct_whitelister_element_rules` hook is tightly bound to the current implementation (e.g. passing `allow_without_attributes` / `attribute_rule` around) and changing it would break existing user code, and 2) draftail _isn't_ going to use it, so this is effectively a legacy code path.

**Comment by gasman** [comment]
Now that I've reached the equivalent point in the Draftail integration work (namely, making `FeatureRegistry` the global source of truth about contentstate/dbHTML conversion rules so that non-core modules can push their own rules into the mix), it looks like we could do with a bit more refactoring to make `FeatureRegistry` less tied to the editorHTML implementation.

Outline plan for doing that:

* Introduce two methods `FeatureRegistry.register_link_type(link_type, handler)` and `FeatureRegistry.register_embed_type(embed_type, handler)` that _just_ deal with the db-to-frontend rewriting. These methods replace `register_embed_handler_rules` and `register_link_handler_rules`.
    * The `handler` for a link_type is called when we encounter an `<a linktype="foo">` opening tag; it takes the dict of attributes and returns the `<a href="...">` tag (as a string) to replace it with.
    * The `handler` for an embed_type is called when we encounter an `<embed embedtype="foo" />` element; it takes the dict of attributes and returns the HTML fragment (as a string) to replace it with
* The db-to-frontend rewriting process isn't aware of feature lists - it will rewrite all recognised linktypes/embedtypes in a rich text value regardless of what the corresponding RichTextField claims it should contain - so we can cut down `get_embed_handler_rules` / `get_all_embed_handler_rules` / `get_link_handler_rules` / `get_all_link_handler_rules` to just `get_link_types()` and `get_embed_types()`.
* Everything to do with editorHTML<->dbHTML and contentState<->dbHTML conversion is condensed into a pair of methods: `FeatureRegistry.register_converter_rule(converter_name, feature_name, rule)` and `FeatureRegistry.get_converter_rule(converter_name, feature_name)`, which work analogously to `register_editor_plugin` / `get_editor_plugin`:
    * just as we have two built-in editor backends named `hallo` and `draftail` (with the potential for third-party apps to add more), we have two converter backends `editorhtml` and `contentstate`
    * the `rule` object passed to `register_converter_rule` is a black box from FeatureRegistry's point of view and is custom to that specific converter backend - just like `HalloPlugin`s are opaque objects only understood by `HalloRichTextArea`.
* `EditorHTMLConverter` will thus retrieve all of its configuration by calling `features.get_converter_rule('editorhtml', some_feature)`, rather than each bit of EditorHTMLConverter logic (whitelisting, linktype conversion in both directions, embedtype conversion in both directions) having its own dedicated method in FeatureRegistry
* This means that EditorHTMLConverter's `rule` objects will have to identify themselves as a whitelisting rule, a linktype rule or an embedtype rule, so that when the EditorHTMLConverter object is being built, they can be directed to the appropriate part of EditorHTMLConverter's implementation.

**Comment by gasman** [comment]
Rebased and updated to make the FeatureRegistry API cleaner, as outlined above.

File: wagtail/core/rich_text/rewriters.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        try:
+            link_type = attrs['linktype']
+        except KeyError:
+            # return ordinary links without a linktype unchanged
+            return match.group(0)
+
+        try:
+            rule = self.link_rules[link_type]
+        except KeyError:
+            # unrecognised link type
```
Comment (BertrandBordage) [inline wagtail/core/rich_text/rewriters.py]: Maybe warn or log this?

File: wagtail/core/rich_text/rewriters.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            return '<a>'
+
+        return rule(attrs)
+
+    def __call__(self, html):
+        return FIND_A_TAG.sub(self.replace_tag, html)
+
+
+class HTMLRewriter:
```
Comment (BertrandBordage) [inline wagtail/core/rich_text/rewriters.py]: The name feels a bit weird, as its a rewriters container.
Maybe just `HTMLRewriters` to make it clearer?

File: wagtail/admin/rich_text/converters/editor_html.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+class EditorHTMLConverter():
+    def __init__(self, features=None):
+        if features is None:
+            features = feature_registry.get_default_features()
+
+        self.converter_rules = []
+        for feature in features:
+            rule = feature_registry.get_converter_rule('editorhtml', feature)
+            if rule is not None:
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/admin/rich_text/converters/editor_html.py]: Please see my answer at https://github.com/wagtail/wagtail/pull/4136#discussion_r158197142.

File: wagtail/admin/rich_text/converters/editor_html.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                tag.unwrap()
+                return
+
+            link_attrs = link_handler.get_db_attributes(tag)
+            link_attrs['linktype'] = link_type
+            tag.attrs.clear()
+            tag.attrs.update(**link_attrs)
+        else:
+            if tag.name == 'div':
+                tag.name = 'p'
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/admin/rich_text/converters/editor_html.py]: I think so, yep. That line has been in place since pre- Wagtail 0.1 :-)

File: wagtail/core/rich_text/__init__.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    global FRONTEND_REWRITER
+
+    if FRONTEND_REWRITER is None:
+        embed_rules = features.get_embed_types()
+        link_rules = features.get_link_types()
+        FRONTEND_REWRITER = HTMLRewriter([
+            LinkRewriter(link_rules), EmbedRewriter(embed_rules)
+        ])
+
+    return FRONTEND_REWRITER(html)
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/core/rich_text/__init__.py]: The way I see it, `FeatureRegistry`'s responsibility is to collect together all the bits of configuration necessary for classes like `EditorHTMLConverter` to do their job. `FeatureRegistry` itself doesn't get involved in performing that conversion task - the rule / plugin objects are "black boxes" from `FeatureRegistry`'s perspective.

HTML rewriting for the front-end is a special case that doesn't fit nicely into that pattern (there's one global rewriter that's always active regardless of which rich text editor is in use, so we need to give `link_types` and `embed_types` [a proper documented spec within FeatureRegistry](https://github.com/gasman/wagtail/blob/9f48f4b9d78e4df1732361cc4cc74d55b31aa5a6/wagtail/core/rich_text/feature_registry.py#L29-L38) rather than saying "it's a black box") - but I think it makes sense to try and stick to the principle that FeatureRegistry doesn't have responsibility for the actual rewriting.

In other words: I'd say that FeatureRegistry is not the right place to implement the logic "an HTMLRewriter consists of a LinkRewriter plus an EmbedRewriter", even if all the individual components do exist there.

File: wagtail/core/rich_text/__init__.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    A custom object used to represent a renderable rich text value.
+    Provides a 'source' property to access the original source code,
+    and renders to the front-end HTML rendering.
+    Used as the native value of a wagtailcore.blocks.field_block.RichTextBlock.
+    """
+    def __init__(self, source):
+        self.source = (source or '')
+
+    def __html__(self):
+        return '<div class="rich-text">' + expand_db_html(self.source) + '</div>'
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/core/rich_text/__init__.py]: Yep, this is a limitation of the rich text implementation as it exists right now. The most immediate blocker is that RichTextField still relies on the `|richtext` template filter, which receives the value as a plain string and thus has no way to trace it back to a feature list. I can see a route to fixing that (starting with phasing out `|richtext`, in favour of `RichTextField` returning a `RichText object) but that's definitely out of scope for this PR.

(I don't think it's worth adding a `# FIXME` - if this is indeed a problem in the real world, then it will come up as a Github issue soon enough. And if it doesn't, then there's no harm in leaving it unfixed :-) )

File: wagtail/core/rich_text/feature_registry.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        plugins[feature_name] = plugin
+
+    def get_editor_plugin(self, editor_name, feature_name):
+        if not self.has_scanned_for_features:
+            self._scan_for_features()
+
+        try:
+            return self.plugins_by_editor[editor_name][feature_name]
+        except KeyError:
+            return None
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/core/rich_text/feature_registry.py]: `self.plugins_by_editor[editor_name].get(feature_name)` would throw a KeyError if `editor_name` is unrecognised - logically this case should return `None` too, so that the outcome of `get_editor_plugin('some_editor', 'unknown_feature')` doesn't vary depending on whether any plugins have been registered against `some_editor`. Admittedly, it's hard to come up with any real-world scenarios where this would actually matter... :-)

[the remark this answers is not in the document]
Comment (BertrandBordage) [reply wagtail/core/rich_text/feature_registry.py]: Ah, right, I thought `editor_name` was always in `self.plugins_by_editor` :)

File: wagtail/admin/rich_text/converters/editor_html.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        return link_handlers
+
+    def clean_tag_node(self, doc, tag):
+        if 'data-embedtype' in tag.attrs:
+            embed_type = tag['data-embedtype']
+            # fetch the appropriate embed handler for this embedtype
+            try:
+                embed_handler = self.embed_handlers[embed_type]
+            except KeyError:
+                # discard embeds with unrecognised embedtypes
```
[the remark this answers is not in the document]
Comment (gasman) [reply wagtail/admin/rich_text/converters/editor_html.py]: I'm undecided about how severe we want to make this error... I suspect we don't want it to fail _loudly_ (because it would affect the site front-end too), and a log message / warning seems too easily ignored to be worth adding. Given that uncertainty, I'd prefer to keep the existing behaviour (silently failing).

File: wagtail/core/rich_text/rewriters.py (hunk 2 above)
Comment (gasman) [reply wagtail/core/rich_text/rewriters.py]: Makes sense - will rename to `MultiRuleRewriter` (I think that's neater than `Rewriters` plural, since it's exposing a single callable like the other rewriters).

**Comment by gasman** [comment]
Thanks for the review @BertrandBordage! Re-pushed with (hopefully) all your feedback accounted for now...

**Comment by gasman** [comment]
Merged in 3da8dce3fc1d8ca27e7bf12f51e86c3193fc9c39 + parents.
